### PR TITLE
Use Minecraft's system for recipes

### PR DESCRIPTION
**Pull request by ramidzkh**
Using Minecraft's system for recipes ensures easy syncing and for players to easily edit recipes through datapacks

---
- [X] Migrate from Ore Dictionary in favor of tags
- [X] Use Minecraft's JSON recipe system

### CODE REVIEW

File: src/main/java/forestry/factory/recipes/CarpenterRecipe.java [hunk 1]
```diff
@@ -55,4 +64,28 @@ public FluidStack getFluidResource() {
 	public IDescriptiveRecipe getCraftingGridRecipe() {
 		return internal;
 	}
+
+	@Override
+	public ResourceLocation getId() {
+		return id;
+	}
+
+	// TODO: Remove ShapedRecipeCustom. We cannot serialize this without ShapedRecipeCustom having a serializer
```
Comment (Nedelosk) [inline src/main/java/forestry/factory/recipes/CarpenterRecipe.java]: Is ShapedRecipeCustom still needed at all anymore?

Comment (ramidzkh) [reply src/main/java/forestry/factory/recipes/CarpenterRecipe.java]: I'm pretty sure it's not, so I'll start removing it

File: src/main/java/forestry/api/recipes/ICarpenterManager.java [hunk 2]
```diff
@@ -54,4 +57,16 @@
...
+
+	/**
+	 * Finds the matching recipe
+	 *
+	 * @param liquid            Present liquid
+	 * @param item              Present item
+	 * @param CraftingInventory Present inventory
+	 * @return An optional carpenter recipe if any matches
+	 */
+	Optional<ICarpenterRecipe> findMatchingRecipe(RecipeManager recipeManager, FluidStack liquid, ItemStack item, IInventory CraftingInventory);
```
Comment (Nedelosk) [inline src/main/java/forestry/api/recipes/ICarpenterManager.java]:
```suggestion
	Optional<ICarpenterRecipe> findMatchingRecipe(RecipeManager recipeManager, FluidStack liquid, ItemStack item, IInventory craftingInventory);
```

File: src/main/java/forestry/factory/recipes/RecipeSerializers.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+		object.addProperty("amount", fluid.getAmount());
+
+		if (fluid.hasTag()) {
+			object.add("Tag", Dynamic.convert(NBTDynamicOps.INSTANCE, JsonOps.INSTANCE, fluid.getTag()));
+		}
+
+		return object;
+	}
+
+	public static JsonObject item(ItemStack stack) {
```
Comment (Nedelosk) [inline src/main/java/forestry/factory/recipes/RecipeSerializers.java:64]: Is there no vanilla / forge method for this?

File: src/main/java/forestry/factory/recipes/RecipeSerializers.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+		FluidStack stack = new FluidStack(fluid, JSONUtils.getInt(object, "amount", 0));
+
+		if (object.has("tag")) {
+			stack.setTag((CompoundNBT) Dynamic.convert(JsonOps.INSTANCE, NBTDynamicOps.INSTANCE, object.get("tag")).copy());
+		}
+
+		return stack;
+	}
+
+	public static JsonObject serializeFluid(FluidStack fluid) {
```
Comment (Nedelosk) [inline src/main/java/forestry/factory/recipes/RecipeSerializers.java:56]: Is there no vanilla / forge method for this?

File: src/main/java/forestry/api/recipes/IForestryRecipe.java [hunk 5]
```diff
@@ -5,5 +5,86 @@
...
+	@Deprecated
+	@Override
+	default boolean isDynamic() {
+		return true;
+	}
+
+	@Deprecated
+	@Override
+	default String getGroup() {
+		return "Forestry";
```
Comment (Nedelosk) [inline src/main/java/forestry/api/recipes/IForestryRecipe.java]: Are all recipes groups uppercase?

**Review by Nedelosk — CHANGES_REQUESTED** [review]

File: src/main/java/forestry/api/recipes/IForestryRecipe.java (hunk 5 above)
Comment (ramidzkh) [reply src/main/java/forestry/api/recipes/IForestryRecipe.java]: Just had a proper look; they are "" if there is no group, or the group is picked from the recipe JSON

File: src/main/java/forestry/factory/recipes/RecipeSerializers.java (hunk 3 above)
Comment (ramidzkh) [reply src/main/java/forestry/factory/recipes/RecipeSerializers.java:64]: Apparently not. I first thought of using Dynamics to convert the NBT item to a JSON object but I don't think ShapedRecipe.deserializeItem would work with it

File: src/main/java/forestry/factory/recipes/RecipeSerializers.java (hunk 4 above)
Comment (ramidzkh) [reply src/main/java/forestry/factory/recipes/RecipeSerializers.java:56]: I couldn't find anything; only for serializing to NBT writing to packet buffers

File: src/main/java/forestry/api/recipes/IForestryRecipe.java (hunk 5 above)
Comment (Nedelosk) [reply src/main/java/forestry/api/recipes/IForestryRecipe.java]: yea, but all the group names from the jsons are lowercase

**Review by Nedelosk — CHANGES_REQUESTED** [review]

**Comment by Nedelosk** [comment]
Looks very nice!

**Comment by ramidzkh** [comment]
All that's left to fix are carpenters, fabricators and worktables. No recipes have actually been converted to JSON (through data generation) yet too

**Comment by ramidzkh** [comment]
All JSON defined recipes should work now. Some code defined still won't work (fabricators and carpenters). This means all that's left is converting the code recipes to use recipe builders (which are not yet implemented). Other than that, I think this is ready for review

File: src/main/java/forestry/factory/tiles/TileCarpenter.java [hunk 6]
```diff
@@ -202,7 +200,7 @@ private boolean removeItemResources(boolean doRemove) {
 
 		NonNullList<ItemStack> craftingSets = InventoryUtil.getStacks(craftingInventory, InventoryGhostCrafting.SLOT_CRAFTING_1, InventoryGhostCrafting.SLOT_CRAFTING_COUNT);
 		IInventory inventory = new InventoryMapper(getInternalInventory(), InventoryCarpenter.SLOT_INVENTORY_1, InventoryCarpenter.SLOT_INVENTORY_COUNT);
-		return InventoryUtil.removeSets(inventory, 1, craftingSets, oreDicts, null, true, false, doRemove);
+		return InventoryUtil.removeSets(inventory, 1, craftingSets, null, true, false, doRemove);
```
Comment (Nedelosk) [inline src/main/java/forestry/factory/tiles/TileCarpenter.java]: Whats about using the ingridient version her too ?

File: src/main/java/forestry/factory/tiles/TileFabricator.java [hunk 7]
```diff
@@ -228,9 +227,9 @@ private void craftResult() {
 		}
 	}
 
-	private boolean removeFromInventory(NonNullList<ItemStack> set, RecipePair<IFabricatorRecipe> recipePair, boolean doRemove) {
+	private boolean removeFromInventory(NonNullList<ItemStack> set, boolean doRemove) {
 		IInventory inventory = new InventoryMapper(this, InventoryFabricator.SLOT_INVENTORY_1, InventoryFabricator.SLOT_INVENTORY_COUNT);
-		return InventoryUtil.removeSets(inventory, 1, set, recipePair.getOreDictEntries(), null, true, false, doRemove);
+		return InventoryUtil.removeSets(inventory, 1, set, null, true, false, doRemove);
```
Comment (Nedelosk) [inline src/main/java/forestry/factory/tiles/TileFabricator.java]: Same like the carpenter

File: src/main/java/forestry/factory/tiles/TileSqueezer.java [hunk 8]
```diff
@@ -161,18 +161,40 @@ public boolean workCycle() {
...
 	private boolean checkRecipe() {
 		ISqueezerRecipe matchingRecipe = null;
+
 		if (inventory.hasResources()) {
 			NonNullList<ItemStack> resources = inventory.getResources();
 
-			if (currentRecipe != null && ItemStackUtil.containsSets(currentRecipe.getResources(), resources, true, false) > 0) {
+			boolean containsSets = false;
+
+			if (currentRecipe != null) {
```
Comment (Nedelosk) [inline src/main/java/forestry/factory/tiles/TileSqueezer.java:172]: Should be replaced with  simulated "consumeIngredients" I thing

**Review by Nedelosk — CHANGES_REQUESTED** [review]

**Comment by ramidzkh** [comment]
I think everything is done except the recipes themselves 🎉

**Review by Nedelosk — APPROVED** [review]

**Comment by ramidzkh** [comment]
Everything seems to be complete now.
---
I need people to playtest it and try to find any bugs relating to the recipes themselves (machines not accepting a certain recipe, the recipes not completing correctly, etc)

Minecraft 1.16.1
Forge `1.16.1-32.0.63`
Download: [Extract this ZIP for the mod jar](https://github.com/ForestryMC/ForestryMC/files/6040775/forestry_extract_me.zip)

**Pull request merged**